### ISSUE TITLE
Know when song is finished with app in background

### ISSUE THREAD

**Issue by YJArc** [spotify/android-sdk#363]
With the current API is there a way to know when the current track has finished when the app is in the background?
The NotificationCallbacks (onPlaybackError, onPlaybackEvent) don't seem to trigger when the app is in background. Would putting the SpotifyPlayer in a service work out this issue? I assumed the player was already a service since it does play in background and use the queue() but currently I can't do anything when a song ends only the API can.
Open to suggestions for work arounds as well.

**Comment by frel** [comment]
You should use a foreground service when playing media. Have a look at how Spotify works. When you press the home button while playing music, a foreground service is started and a notification is shown that you cannot dismiss (like you normally can). This will ensure you that Android does not kill your app.

**Comment by YJArc** [comment]
Thanks for the info @frel
Yes, as I kept coding I was finding more issues with putting it in a service, I do not recommend.
I am still getting familiar with android.
Do you have any recommended resources on doing just that? Getting the app to run while in background.

**Comment by frel** [comment]
I would say on the contrary, put it in a bound service (https://developer.android.com/guide/components/bound-services.html) and when your app is minimized make it run as a foreground service (https://developer.android.com/guide/components/services.html#Foreground).

In Android any application that is minimized will eventually be killed by the system. This happens to free resources. The only exception are foreground applications (The application you are using) and foreground services. If you want to play music in the background (or do some long living important task like file syncing/downloading) you should use a foreground service. Also be aware that the use of regular background services are being limited in Android O (https://developer.android.com/preview/behavior-changes.html#atap).

**Comment by seahorsepip** [comment]
Maybe a MediaBrowserService (extended foreground service) seems like a thing to use? I'm planning of implementing spotify using the same architecture as the google music player sample project: https://github.com/googlesamples/android-UniversalMusicPlayer

A video explaining the classes used in the sample: https://www.youtube.com/watch?v=XQwe30cZffg

**Comment by YJArc** [comment]
Right, so service must be used. I realized it when you close the player the system wouldn't run the callbacks. It just seems like a lot of work for something that seems simple.

Interesting videos. I very much look forward to the seeing your project @seahorsepip
In the mean time, I'll see how far I can go.
